File: src/pipeline/emit/emitter.ts
import type { SerializedIR } from "../../ir/serialized/serialized-ir.js";
import type { EmitOptions } from "./emit-options.js";

export type Emitter = Readonly<{
  format: string;
  contentType: string;
  // File extension without leading dot, used by `--out-dir` to derive
  // output filenames. Multiple emitters may share an extension (e.g. ir
  // and json both write JSON).
  extension: string;
  emit(ir: SerializedIR, opts: EmitOptions): string;
}>;

File: crates/unsnarl-emitter/src/emitter.rs
//! Emitter trait.
//!
//! `format`, `content_type`, and `extension` are returned by `&self`
//! getters so the trait stays object-safe (`&dyn Emitter`) for the
//! registry.

use unsnarl_ir::serialized::SerializedIR;

use crate::emit_options::EmitOptions;

pub trait Emitter {
    fn format(&self) -> &'static str;
    fn content_type(&self) -> &'static str;
    fn extension(&self) -> &'static str;
    fn emit(&self, ir: &SerializedIR, opts: &EmitOptions) -> String;
}
